### DISCUSSION TITLE
Does not work with Laravel Mix

### DISCUSSION THREAD

**Discussion by loilo** [alpinejs/alpine#2454] [1. Help]
I've gone through quite a rabbit hole after copying some Alpine code to a new project that previously worked with the same Laravel Mix based build chain just a few months ago – the only change being that I'm running the new project on Alpine 3.7 while the old one had 3.4 installed.

In the end, the gist is this: c5ffa11a attempted to improve error resiliency but introduced a tiny new flaw.

A little more detail:

Expressions (like `x-data` values) are converted to async functions via the `generateFunctionFromString` function. That function uses the `AsyncFunction` constructor to achieve this.
However, in environments that use Babel to the max (like Laravel Mix, which has `@babel/preset-env`'s `forceAllTransforms` option on by default) the `AsyncFunction` constructor does _not_ actually create an async function, it will only just create a regular function.

This is not a problem unless you use `await` in the expression.

However, said commit attached a `.catch()` to the expression function call, (accidentally) ensuring that only a _real_ async function can be be used. 

Although Alpine is not supposed to run in legacy browsers which run only ES5, this should probably be fixed as devs may not always be aware how their build tools work exactly. (At least I wasn't aware that Mix does only partly respect my `.babelrc` file and applies transforms anyway, after one year of using it.)

This should be an easy fix — I think you could just wrap [the expression function call](https://github.com/alpinejs/alpine/blob/3c578e5c8df402a49c8d84377b586380231d0500/packages/alpinejs/src/evaluator.js#L94) in a `Promise.resolve()` and call it a day. However, I'd like to hear other opinions before possibly submitting a PR.

**Comment by SimoTod** [comment]
I think you should raise that PR, loads of people on discord have failing builds because many bundlers target es2016 by default and the requirement is not mentioned anywhere in the docs.

**Reply by loilo** [reply]
Might take a moment — the fix is unfortunately not _as_ straightforward as I hoped, but I'm on it.

**Reply by loilo** [reply]
See https://github.com/alpinejs/alpine/pull/2467

**Reply by SimoTod** [reply]
the use of eval is a bit controversial. I think Caleb wanted to avoid it so Alpine doesn't get labelled as "the library that uses Unsafe eval". I know it's not really parsing user input but people will just complain.
If you want down that route, you probably want to use `const AsyncFunction = new Function("return Object.getPrototypeOf(async function(){}).constructor")();` which is closer to the original implementation (with a comment explaining why we don't do `Object.getPrototypeOf(async function(){}).constructor` directly).

**Reply by loilo** [reply]
Honest question: Do people react differently when somebody uses `new Function()`? 😁

**Reply by SimoTod** [reply]
Well, some do some don't. But it's already used in the framework and it's seen as a slightly more secure version. You can wait for Caleb's feedback anyway, just an option. 👍

**Reply by loilo** [reply]
No I think it's fine, I don't mind. Whatever gets us closer to resolving the issue. 😉

**Comment by mauriciv** [comment]
Hey @loilo did you find a way to make it work with Laravel Mix?

**Reply by loilo** [reply]
Yes, it's possible by telling Mix to not generate ES5 in the first place:

```js
mix.babelConfig({
  presets: [['@babel/preset-env', { modules: false, forceAllTransforms: false }]]
})
```

**Reply by adriansio** [reply]
@loilo Hi, unfortunately I am not able to use Livewire 3, Alpine and Laravel Mix together. I keep getting this error:

```
Alpine Expression Error: func(...).catch is not a function
Uncaught TypeError: func(...).catch is not a function
```

Do I need to have a `.babelrc`?

**Reply by loilo** [reply]
@adriansio You do not _need_ a `.babelrc`, the code I gave above is for the `webpack.mix.js`. However, you _can_ write equivalent configuration in a `.babelrc` like so:

```json
{
  "presets": [
    [
      "@babel/preset-env",
      {
        "modules": false,
        "forceAllTransforms": false
      }
    ]
  ]
}
```

**Reply by adriansio** [reply]
@loilo Thanks but still not working. Were you able to successfully use Livewire 3, Alpine 3 and Laravel Mix with the mix configuration option you posted?
